bin/metababel

```diff
@@ -121,46 +161,20 @@ def wrote_event_dispatchers(folder, trace_model, trace_matching_model)
     dispatcher
   end
 
-  dispatch_types = dispatchers.flat_map(&:dispatch_types)
-  dispatch_types += trace_matching_model.map_event_classes_with_index do |em, _, _, default_clock_class|
-    matched_dispatchers, signatures = dispatchers.filter_map do |dispatcher|
-      extracted_args = dispatcher.event.match?(em)
-      [dispatcher, extracted_args] if extracted_args
-    end.transpose
-
-    if !matched_dispatchers or matched_dispatchers.empty?
-      raise "No event matched '#{em.callback_name}', at least one matching event required."
-    end
-
-    # Verify same signature
-    unique_signatures = signatures.uniq
-    unless unique_signatures.size == 1
-      signatures_str = unique_signatures.map { |s| "(#{s.map(&:type).join(', ')})" }.join(', ')
-      raise "Conflicting signatures for '#{em.callback_name}', found #{unique_signatures.size} signatures, only one allowed: '#{signatures_str}'"
-    end
-
-    # Modify the dispatcher to add the new dispatchType
-    name_sanitized = sanitize(em.callback_name)
-    dispatch_type_args = []
-    if default_clock_class and default_clock_class.fetch(
-      :extract, true
-    )
-      dispatch_type_args += [GeneratedArg.new('int64_t',
-                                              '_timestamp')]
-    end
-    dispatch_type_args += [GeneratedArg.new('const char *', '_event_class_name')]
-    dispatch_type_args += signatures.pop
-    dispatch_type = DispatchType.new(name_sanitized, dispatch_type_args, "matching_#{name_sanitized}",
-                                     matched_dispatchers)
+  automatic_dispatch_types = dispatchers.flat_map(&:dispatch_types)
 
-    matched_dispatchers.each do |dispatcher|
-      dispatcher.dispatch_types << dispatch_type
-    end
+  context = { "all" => dispatchers.to_set }
+  dispatch_types = trace_matching_model.map_event_classes_with_index do |em, _, _, default_clock_class|
+    matched_dispatchers = em.domain ? OpenStruct.new(context).instance_eval(em.domain) : context["all"]
+    raise "Nil or empty event set for '#{em.set_id}'." unless matched_dispatchers.is_a?(Set)
 
-    dispatch_type
-  end
+    # We mutate matched_dispatchers, context.
+    get_dispatch_type(em, default_clock_class, matched_dispatchers, context)
+  end.compact
 
-  d = { dispatchers: dispatchers, dispatch_types: dispatch_types, event_name: event_name }
+  d = { dispatchers: dispatchers,
+        dispatch_types: automatic_dispatch_types + dispatch_types,
```

Comment:
Maybe rename `dispatch_types` to `matching_dispatch_types` ? Just to be a little clearer